### PR TITLE
Ensure anonymous schemas have an empty names list

### PR DESCRIPTION
**Pull request by alxrsngrtn**
- Added parameterized tests for `isMoreSpecific`
- Added tests for parsing and constructing anonymous schemas
- Added tests + a fix to ensure that bad anonymous schemas don't accidentally get created.

### CODE REVIEW

**Comment by alxrsngrtn** [comment]
... Eh. This PR is not necessary anymore. 

I introduced a problem in a separate PR that this was supposed to fix. The long-story-short of why it was hard to trace: a readonly property was getting modified outside of the class through references.

```
const keys = schema.names
// ...
keys.push('')
```
I though this would add the empty string to a _copy_ of the data. Instead it was modifying the class member. 

Anyway, I can clean up this PR tomorrow -- it adds tests of schema functionality, which is not a bad thing. I can revert the change to schema's behavior, too.

File: src/runtime/schema.ts [hunk 1]
```diff
@@ -30,7 +30,7 @@ export class Schema {
   // in `fields`; the constructor will convert these to the correct schema form.
   // tslint:disable-next-line: no-any
   constructor(names: string[], fields: Dictionary<any>, description?) {
-    this.names = names;
+    this.names = names.filter((value => !!value));
```
Comment (piotrswigon) [inline src/runtime/schema.ts]: You have fixed the symptom but not the underlying issue. Why was the list of names `[undefined]` in the first place?

It is true that passing a an array with undefined, null or an empty string is wrong here, but we cannot do full validation in every constructor and in every method - that would be too costly.

It is best to fix the problem at its root, and the actual bug is in processing of the parsed output in `manifest.ts`. The `_processSchema` method has such a funky line in it:

`names = [names[0], ...names.filter(name => name !== names[0])];`

This looks to me like a source of our pain. A name array can be empty, but after this line it will be a one element array of undefined.

What this line attempts to achieve (I think) is avoiding duplicate names I presume, I would replace it with:

`names = [...new Set(names)];`

File: src/runtime/tests/schema-test.ts [hunk 2]
```diff
@@ -443,4 +444,86 @@ describe('schema', () => {
...
+
+  const notPartialOrdered = (testCase) => {
+    const [caseName, first, second] = testCase;
+    it(`determines entities have no overlap: ${caseName}`, () => {
+      assert.isFalse(first.isMoreSpecificThan(second));
+      assert.isFalse(second.isMoreSpecificThan(first));
+    });
+  };
+
+  const partiallyOrderedTestCases = async () => {
```
Comment (piotrswigon) [inline src/runtime/tests/schema-test.ts]: These tests have little to do with the bug you are fixing, please remove them from this patch. If you are testing `isMoreSpecificThan` please do it in a separate PR.

File: src/runtime/tests/schema-test.ts [hunk 3]
```diff
@@ -443,4 +444,86 @@ describe('schema', () => {
...
+      out * {Number n, Bytes b} output2`);
+
+    const particle = manifest.particles[0];
+    const schema = (conn: HandleConnectionSpec): Schema => conn.type.getEntitySchema();
+    particle.connections
+      .map(schema)
+      .forEach((s: Schema) => assert.deepEqual(s.names, []));
+  });
+
+  function testPoorAnonConstruction() {
```
Comment (piotrswigon) [inline src/runtime/tests/schema-test.ts]: This test is complex and I have to warm up my brain to understand it - that is not good. Tests should be trivial if possible.

What you want, once you fix the problem at the root is something like this:

```
  it('parses annonymous schemas', async () => {
    const manifest = await Manifest.parse(`alias schema * as X`);
    const schema = manifest.findSchemaByName('X');

    assert.isEmpty(schema.names);
    assert.isEmpty(schema.fields);
  });

  it('parses annonymous inline schemas', async () => {
    const manifest = await Manifest.parse(`
      particle P
        in * {} thing`);

    const particle = manifest.particles[0];
    const connection = particle.handleConnections[0];
    const schema = connection.type.getEntitySchema();

    assert.isEmpty(schema.names);
    assert.isEmpty(schema.fields);
  });
```

Sorry for doing your job for you, but sometimes it is easier to show with an example rather than talk abstractly.

Do you see how simple these tests are? There is no iteration over collection to then programmatically generate inputs and tests cases using them. This is what you should strive for - code that is as simple and as readable and as focused as you can make it.

**Comment by piotrswigon** [comment]
Ok, I see you want to close this PR, I will send you a fix to the underlying problem with undefined.

File: src/runtime/schema.ts (hunk 1 above)
Comment (alxrsngrtn) [reply src/runtime/schema.ts]: According to some of my other tests, this does not cause the problem. I traced the issue I was experiencing to code I introduced in another PR. 

This suggestion is a valid improvement.

File: src/runtime/tests/schema-test.ts (hunk 2 above)
Comment (alxrsngrtn) [reply src/runtime/tests/schema-test.ts]: Ok. I'll move them to another PR. I'll clean them up first.

**Comment by alxrsngrtn** [comment]
PR not necessary anymore, functionality implemented elsewhere.
